README.md
# PIMPLE

**P**erformance **I**ndicators for **M**anagement **P**rocedures exp**L**or**E**r

An R-Shiny tool for exploring the MSE results for WCPFC harvest strategies.

## Note

The code for older versions of PIMPLE (up to 10/10/2022) are in the old [ofp-sam-amped](https://github.com/PacificCommunity/ofp-sam-amped) repository.

The old repository has all the AMPLE apps and PIMPLE. These projects are no longer linked.
It became awkward to individually maintain these distinct projects and their releases with a single repo.

Hence, this repo, started on the 10/10/2022, has just PIMPLE. It takes the PIMPLE code from the *PIMPLE_SMD_2022* branch of the old repo.


introtext/news.md
# News

## December 2022

* Added adopted HCR (MP) adopted by WCPFC19.
* Only Adopted HCR is selected at the start.

## November 2022

* Added robustness sets including hyperstability; low recruitment; and effort creep scenarios.
* Changed the SB/SBF=0 relative to SB/SBF=0 in 2012 indicator to SB/SBF=0 relative to target (where the target is determined as equilibrium stock status under baseline fishing levels).
* Added effort relative to effort in reference period indicator for effort driven fisheries (purse seine and pole & line).
* Box plots are now the default presentation method instead of bar charts.
* Moved pole and line CPUE to 'Compare performance' tab.

## October 2022

* Release of PIMPLE for MOC in preparation for WCPFC 19 (https://ofp-sam.shinyapps.io/PIMPLE_WCPFC19/)
* Following guidance from the WCPFC first Science-Management Dialogue meeting, the HCR selection has been reduced to HCR 1, 2, 5, 6 and 9, all with a +-10% constraint on how much the output of the HCR can change between management periods.
* Pole and line fisheries now managed through setting of effort limits (relative to the effort 2001-2004) instead of catch.
* Added pole and line CPUE, relative to 2001-2004 under the 'Other SKJ indicators' tab.
* Added HCR parameters to the 'SKJ management procedures' tab.





funcs.R
# General functions

#' SPC about information
#' 
#' SPC about information included in all the Shiny apps.
#' @noRd
#' @keywords internal
spc_about <- function(){
  out <- tags$html(
    tags$p(style="opacity: 0.5", class="caption", align="center", HTML("&copy"), "Pacific Community, 2021"),
    tags$h1("About us:"),
    tags$p(align="justify", "The Pacific Community (SPC) is the principal scientific and technical organisation in the Pacific region, proudly supporting development since 1947. It is an international development organisation owned and governed by its 26 country and territory members. The members are: American Samoa, Australia, Cook Islands, Federated States of Micronesia, Fiji, France, French Polynesia, Guam, Kiribati, Marshall Islands, Nauru, New Caledonia, New Zealand, Niue, Northern Mariana Islands, Palau, Papua New Guinea, Pitcairn Islands, Samoa, Solomon Islands, Tokelau, Tonga, Tuvalu, United States of America, Vanuatu, and Wallis and Futuna."),
    tags$p(align="justify", "In pursuit of sustainable development to benefit Pacific people, this unique organisation works across more than 25 sectors. SPC is renowned for its knowledge and innovation in such areas as fisheries science, public health surveillance, geoscience and conservation of plant genetic resources for food and agriculture."),
    tags$p(align="justify", "Much of SPC's focus is on major cross-cutting issues, such as climate change, disaster risk management, food security, gender equality, human rights, non-communicable diseases and youth employment. Using a multi-sector approach in responding to its members' development priorities, SPC draws on skills and capabilities from around the region and internationally, and supports the empowerment of Pacific communities and sharing of expertise and skills between countries and territories."),
    tags$p(align="justify", "With over 600 staff, SPC has its headquarters in Noumea, regional offices in Suva and Pohnpei, a country office in Honiara and field staff in other Pacific locations. Its working languages are English and French. See: ", a("https://www.spc.int", href="www.spc.int"))
  )
}

# The big PI tables for comparing HCRs
#' pitable
#'
#' pitable() is not a plot but a table comparing PIs across HCRs and periods. Only pass in 1 time period at a time.
#'
#' @param signif Number of significent digits for table. Default is 3.
#'
#' @return A data.frame to be shown as a table.
#' @rdname comparison_plots
#' @name Comparison plots
#' @export
# Added signif argument to be added to AMPLE
pitable <- function(dat, percentile_range = c(10,90), signif=3){
  # Rows are the PIs, columns are the HCRs
  colmin <- paste("X",percentile_range[1],".",sep="")
  colmax <- paste("X",percentile_range[2],".",sep="")
  percentile_min <- dat[, get(colmin)]
  percentile_max <- dat[, get(colmax)]
  dat$value <- paste(signif(dat$X50.,signif), " (", signif(percentile_min, signif), ",", signif(percentile_max, signif), ")", sep="")
  # Fix pi1
  dat[dat$pi=="pi1", "value"] <- signif(dat[dat$pi=="pi1", "X50."],signif)
  tabdat <- dat[,c("hcrref", "piname", "value")]
  tabdat[tabdat$name=="Biomass","piname"] <- "SB/SBF=0"
  #tabdat <- as.data.frame(spread(tabdat, key="hcrref", value="value"))
  tabdat <- as.data.frame(dcast(tabdat, hcrref ~ piname, value.var = "value"))
  # Have rownames?
  #rnames <- tabdat[,1]
  #tabdat <- tabdat[,-1]
  #rownames(tabdat) <- rnames
  colnames(tabdat)[1] <- "Indicator"
  return(tabdat)
}

# Add or remove HCRs as selected? Sure
get_hcr_param_table <- function(){
  # These data should be produced during the PI calculation
  # But for we now we can hardwire it
  # Ordering the names and empty rows
  all_hcr_names <-  c("Adopted WCPFC19", "HCR 1 (+-10%)", "HCR 2 (+-10%)", "HCR 5 (+-10%)", "HCR 6 (+-10%)", "HCR 9 (+-10%)", "HCR 9b (+-10%)", "HCR 9c (+-10%)", "HCR 9d (+-10%)") 
  all_hcr_names_padding <- c(rbind(all_hcr_names, rep(" ", length(all_hcr_names))))
  
  hcrtab <- data.frame(hcr_name = rep(all_hcr_names, each=2),
             hcr = all_hcr_names_padding,
             metric = c("SB/SBF=0", "HCR output"),
             limit = 0.2,
             threshold = c(0.4, 1.0, 0.42, 1.0, 0.42, 1.0, 0.37, 1.0, 0.37, 1.0, 0.37, 1.0, 0.37, 1.0, 0.37, 1.0, 0.37, 1.0),
             step_end = c(0.57, 1.0, rep(" ", 6), 0.47, 1.0, 0.47, 1.0, 0.47, 1.0, 0.47, 1.0, 0.57, 1.0),
             maximum = c(0.8, 1.2, rep(" ", 6), 0.9, 1.3, 0.8, 1.4, 0.9, 1.2, 0.8, 1.2, 0.8, 1.2))
  # Fix column names
  colnames(hcrtab) <- c("hcr_ref", "HCR", " ", "Limit", "Threshold", "Step end", "Maximum")
  #hcrtab <- hcrtab[hcrtab$hcr_name %in% hcr_choices,]
  return(hcrtab)
}



plots.R
get_hcr_colours <- function(hcr_names, chosen_hcr_names){
  allcols <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(12,"Paired"))(length(hcr_names))
  names(allcols) <- hcr_names
  hcrcols <- allcols[as.character(chosen_hcr_names)]
  return(hcrcols)
}

get_betmp_colours <- function(mp_names, chosen_mp_names){
  allcols <- grDevices::colorRampPalette(RColorBrewer::brewer.pal(9,"Set1"))(length(mp_names))
  names(allcols) <- mp_names
  mpcols <- allcols[as.character(chosen_mp_names)]
  return(mpcols)
}



barboxplot <- function(dat, hcr_choices, plot_type="median_bar", quantiles=c(0.10, 0.25, 0.75, 0.90), no_cols=2){
  all_hcr_names <- unique(dat$hcrref)
  dat <- dat[dat$hcrref %in% hcr_choices,] # Trying to remove weird warning about hcr_no being a global variable
  hcr_cols <- get_hcr_colours(hcr_names=all_hcr_names, chosen_hcr_names=unique(dat$hcrref))
  if (plot_type=="median_bar"){
    #p <- ggplot(dat, aes_string(x="period", y="X50.", fill="hcrref"))
    p <- ggplot(dat, aes(x=period, y=X50., fill=hcrref))
    p <- p + geom_bar(stat="identity", position="dodge", colour="black", width=0.7)
    p <- p + ylab("Average value")
  }
  if (plot_type=="box"){
    quantiles <- sort(quantiles)
    #quantiles_text <- paste("value_", quantiles, sep="") 
    quantiles_text <- paste("X", quantiles * 100, ".", sep="") 
    ymin <- quantiles_text[1]
    ymax <- quantiles_text[4]
    lower <- quantiles_text[2]
    upper <- quantiles_text[3]
    #p <- ggplot(dat, aes_string(x="period"))
    p <- ggplot(dat, aes(x=period))
    #p <- p + geom_boxplot(aes_string(ymin=ymin, ymax=ymax, lower=lower, upper=upper, middle="X50.", fill="hcrref"), stat="identity", width=0.7)
    p <- p + geom_boxplot(aes(ymin=.data[[ymin]], ymax=.data[[ymax]], lower=.data[[lower]], upper=.data[[upper]], middle=X50., fill=hcrref), stat="identity", width=0.7)
  }
  p <- p + xlab("Time period")
  p <- p + scale_fill_manual(values=hcr_cols)
  p <- p + facet_wrap(~piname, scales="free", ncol=no_cols)
  p <- p + theme_bw()
  p <- p + theme(legend.position="top", legend.title=element_blank())
  
  # Size of labels etc
  text_size <- 14
  p <- p + theme(axis.text=element_text(size=text_size), axis.title=element_text(size=text_size), strip.text=element_text(size=text_size), legend.text=element_text(size=text_size))
  return(p)
}

time_series_plot <- function(dat, hcr_choices, wormdat=NULL,
                          linetype_worm=1,
                          outer_percentile_range = c(5,95),
                          inner_percentile_range = c(10,90),
                          colour_worm="black",
                          size_worm=0.3, add_start_line=TRUE, time_period_lines=TRUE,
                          short_term = 2016:2024, medium_term = 2025:2033, long_term = 2034:2042,
                          last_plot_year=2042, show_spaghetti=FALSE){
  # Sort out colours based on chosen HCRs
  all_hcr_names <- unique(dat$hcrref)
  dat <- dat[dat$hcrref %in% hcr_choices,] # Trying to remove weird warning about hcr_no being a global variable
  hcr_cols <- get_hcr_colours(hcr_names=all_hcr_names, chosen_hcr_names=unique(dat$hcrref))
  # Rename for percentiles reference
  inner_ynames <- paste0("X", inner_percentile_range, ".")
  outer_ynames <- paste0("X", outer_percentile_range, ".")
  # Chop out NA rows - better to leave in and plot the gaps
  #dat <- dat[!is.na(dat$X50.),]
  # Start the plot
  p <- ggplot(dat, aes(x=year))
  # The ribbon
  #p <- p + geom_ribbon(aes_string(ymin=outer_ynames[1], ymax=outer_ynames[2], fill="hcrref"), alpha=0.5)
  p <- p + geom_ribbon(aes(ymin=.data[[outer_ynames[1]]], ymax=.data[[outer_ynames[2]]], fill=hcrref), alpha=0.5)
  #p <- p + geom_ribbon(aes_string(ymin=inner_ynames[1], ymax=inner_ynames[2], fill="hcrref"))
  p <- p + geom_ribbon(aes(ymin=.data[[inner_ynames[1]]], ymax=.data[[inner_ynames[2]]], fill=hcrref))
  # Add median line
  #p <- p + geom_line(aes(y=X50., group=hcrref), colour="black", linetype=2, size=rel(0.5))
  p <- p + geom_line(aes(y=X50., group=hcrref), colour="black", linetype=2, linewidth=rel(0.5))
  p <- p + xlab("Year")
  # Add worms
  # These look bad - hard to see the colours
  if (!is.null(wormdat) & show_spaghetti==TRUE){
    wormdat <- subset(wormdat, hcrref %in% hcr_choices)
    #wormdat <- wormdat[!is.na(wormdat$value),]
    # Put a black background on the line to help
    p <- p + geom_line(data=wormdat, aes(x=year, y=value, group=wormid), colour="black", linetype=linetype_worm, linewidth=size_worm*1.2)
  }
  # Colours
  p <- p + scale_fill_manual(values=hcr_cols)
  p <- p + scale_colour_manual(values=hcr_cols)
  # Add vertical line at start of MSE
  if (add_start_line){
    p <- p + geom_vline(aes(xintercept=min(short_term)-3), linetype=2)
  }
  # Time period lines?
  if(time_period_lines){
    p <- p + geom_vline(aes(xintercept=min(short_term)), linetype=2)
    p <- p + geom_vline(aes(xintercept=min(medium_term)), linetype=2)
    p <- p + geom_vline(aes(xintercept=min(long_term)), linetype=2)
  }
  
  # Faceting by PI - for comparing lots of different metrics - add later
  #p <- p + facet_wrap(~piname, scales="free", ncol=1)
  #p <- p + facet_grid(piname ~ hcrref, scales="free")#, ncol=1)
  p <- p + theme_bw()
  
  p <- p + theme(legend.position="top", legend.title=element_blank())
  
  return(p)
}


hcr_plot <- function(hcr_choices, hcr_shape, hcr_points, lrp, trp, add_points=FALSE, add_path=FALSE, blacklinesize=4, linesize=3, pointsize=4.2, stroke=3){
  hcrcols <- get_hcr_colours(hcr_names=unique(hcr_shape$hcrref), chosen_hcr_names=hcr_choices)
  # Select the chosen HCRs only - could do this in the call to plot in app?
  shapedat <- subset(hcr_shape, hcrref %in% hcr_choices)
  pointsdat <- subset(hcr_points, hcrref %in% hcr_choices)
  p <- ggplot(shapedat, aes(x=x, y=y))
  p <- p + geom_line(aes(group=hcrref), colour="black", linewidth=blacklinesize) # outline
  p <- p + geom_line(aes(colour=hcrref), linewidth=linesize)
  p <- p + xlab("Estimated SB/SBF=0") + ylab("Effort multiplier")
  p <- p + theme_bw()
  p <- p + theme(legend.position="top", legend.title=element_blank())
  p <- p + scale_fill_manual(values=hcrcols)
  p <- p + scale_colour_manual(values=hcrcols)
  p <- p + theme(axis.text=element_text(size=16), axis.title=element_text(size=16), strip.text=element_text(size=16), legend.text=element_text(size=16))
  p <- p + scale_x_continuous(expand = c(0, 0))
  # Add points to it
  if (add_points){
    p <- p + geom_jitter(dat=pointsdat, aes(x=sbsbf0, y=scaler, fill=hcrref), width=0.00, height=0.015, colour="black", shape=21, size=4.2, stroke=2, alpha=0.5)
  }
  if (add_path){
    # Connect the iters by lines so you can see what happens
    # No alpha or jittering on these points
    p <- p + geom_point(dat=pointsdat, aes(x=sbsbf0, y=scaler, fill=hcrref), colour="black", shape=21, size=4.2, stroke=2)
    # Add thicker point at start
    minyear <- min(pointsdat$year)
    p <- p + geom_point(dat=pointsdat[year==minyear,], aes(x=sbsbf0, y=scaler, fill=hcrref), colour="black", shape=21, size=3.2, stroke=4)
    p <- p + geom_path(dat=pointsdat, aes(x=sbsbf0, y=scaler, group=interaction(iter, hcrref)), colour="black", linewidth=1.5)
  }
  # Add LRP and TRP
  #p <- p + geom_vline(aes(xintercept=lrp), linetype=2)
  #p <- p + geom_vline(aes(xintercept=trp), linetype=2)
  p <- p + ylim(0, NA)
  return(p)
}

mixpis_barbox_biol_plot <- function(dat, hcr_choices, betmp_choices, barbox_choice = "median_bar", stock_choice = stock_choice, facetskjorbet = "skjhcr", no_mixfacets_row=3){
  
    # SKJ HCR and BET MP colours
    all_hcr_names <- unique(dat$skjhcrref)
    all_betmp_names <- unique(dat$tll_ass_name)
    dat <- dat[dat$hcrref %in% hcr_choices & dat$tll_mult %in% betmp_choices,] 
    hcr_cols <- get_hcr_colours(hcr_names=all_hcr_names, chosen_hcr_names=unique(dat$skjhcrref))
    betmp_cols <- get_betmp_colours(mp_names=all_betmp_names, chosen_mp_names=unique(dat$tll_ass_name))
  
    # Option 1      
    # Each facet is a TLL - 3 time periods in each facet - each bar as a Z - similar to existing
    #fillvar = "hcrref" # Or "tll_ass_name"
    if (facetskjorbet == "skjhcr"){
      fillvar = "tll_ass_name"
    } else {
      fillvar = "skjhcrref"
    }
    p <- ggplot(dat, aes(x=period))
    if (barbox_choice == "box"){
      p <- p + geom_boxplot(aes(ymin=X10., ymax=X90., lower=X25., upper=X75., middle=X50., fill=.data[[fillvar]]), stat="identity")
    }
    if (barbox_choice == "median_bar"){
      p <- p + geom_bar(aes(y=X50., fill=.data[[fillvar]]), stat="identity", position="dodge", colour="black", width=0.7)
    }
    p <- p + xlab("Time period")
    if (fillvar == "skjhcrref"){
      p <- p + scale_fill_manual(values=hcr_cols, name="SKJ HCR")
      p <- p + facet_wrap(~tll_ass_name, ncol=no_mixfacets_row)
    }
    if (fillvar == "tll_ass_name"){
      p <- p + scale_fill_manual(values=betmp_cols, name="TLL scenario")
      p <- p + facet_wrap(~skjhcrref, ncol=no_mixfacets_row)
    }
    p <- p + theme_bw()
    p <- p + theme(legend.position="top")#, legend.title=element_blank())
    p <- p + guides(fill = guide_legend(title.position = "top", title.hjust=0.5))
    
    text_size <- 14
    p <- p + theme(axis.text=element_text(size=text_size), axis.title=element_text(size=text_size), strip.text=element_text(size=text_size), legend.text=element_text(size=text_size))
    
    return(p)
  
  
}


  